### PR TITLE
Method tracers refactored

### LINKED ISSUE #349: Ruby 2.7 deprecation warning tracks the method wrongly to new relic line
I'm not quite sure if this issue belongs here or in Ruby project. 

Ruby tracks the message defined in another library/project to this line on NewRelic gem, https://github.com/newrelic/rpm/blob/main/lib/new_relic/agent/method_tracer.rb#L337, and then throws the deprecation warning.

```ruby
# trigger_profile_confirmed_hooks.rb

class TriggerProfileConfirmedHooks
  Messenger.publish("user.confirmed", {foo: “bar”})
end


# messenger.rb
class Messenger
  def self.publish(message_type, payload = {}, opts = {})
  end
end
```

Warning:

```
app/services/trigger_profile_confirmed_hooks.rb|10 warning| Using the last argument as keyword parameters is deprecated; maybe ** should be added to the call

/.rbenv/versions/2.7.0/lib/ruby/gems/2.7.0/gems/newrelic_rpm-6.11.0.365/lib/new_relic/agent/method_tracer.rb|337 warning| The called method `publish' is defined here
```

Apparently, it happens only for particular method signature. If I change the method name on my project to:
`def something_else(message_type, payload = {}, opts = {})`,
instead of `def publish(message_type, payload = {}, opts = {})`,
the warning goes away.

May be this will go away, when New Relic gem will handle all the 2.7 warnings. Or may be this is related with Ruby wrongly tracing method call.

### LINKED ISSUE #423: ruby 2.7 deprecation warning originating in delayed_job_instrumentation.rb
## Description

Upgrading to ruby 2.7.1, we are seeing the following warning lines numerous times when delayed jobs are run:

```
Sept 06 16:48:32 app/worker.1 /app/vendor/bundle/ruby/2.7.0/gems/delayed_job-4.1.8/lib/delayed/tasks.rb:9: warning: Using the last argument as keyword parameters is deprecated; maybe ** should be added to the call
Sept 06 16:48:32 app/worker.1 /app/vendor/bundle/ruby/2.7.0/gems/newrelic_rpm-6.12.0.367/lib/new_relic/agent/instrumentation/delayed_job_instrumentation.rb:113: warning: The called method `initialize' is defined here
```

## Expected Behavior
There should be no warnings.

### LINKED ISSUE #424: add_method_tracer modifies object arguments
`add_method_tracer` causes object arguments to be cloned.

## Description
When tracing a method that takes an object argument that is supposed to be mutated during the execution of that method, the tracing causes the passed argument to be a clone of the object, meaning that any mutations are lost to outside references to the same object.

## Expected Behavior
No cloning of arguments should happen and the passed object should be included in the call to the method as is.

## Steps to Reproduce
```ruby
class Demonstration
  include ::NewRelic::Agent::MethodTracer

  def call
    (1..10).each_with_object({}) do |i, hash|
      process(i, hash)
    end
  end

  def process(i, hash)
    hash[i] = i * 2
  end

  add_method_tracer :process, 'Demonstration/process'
end

Demonstration.new.call
#=> {}
```

Expected result:

```ruby
Demonstration.new.call
=> {1=>2, 2=>4, 3=>6, 4=>8, 5=>10, 6=>12, 7=>14, 8=>16, 9=>18, 10=>20}
```

## Your Environment
ruby 2.7.1p83 (2020-03-31 revision a0c7c23c9c) [x86_64-darwin19]

## Additional context
It seems like this is not an issue in Ruby 2.6

### PR DESCRIPTION
**Pull request by mwlang**
# Overview
This PR aims to solve some of the 2.7+ issues encountered with method argument delegation when using `add_method_tracer` to instrument specific methods.

# Related Github Issue
* Resolves #424 by not double-splatting `**kwargs`
* Resolves #423 by removing `(*args, **kwargs)` in the tracing method's definition (we were overzealous as with RackApp #486)
* Resolves #349 by not using `(*args, **kwargs)` and by also adding `ruby2_keywords` to identify the method_tracer wrapper as a delegator of method arguments in a way that's compatible with all the rubies.  

NOTE: `ruby2_keywords` expected to be retired in Ruby 3.2, so must refactor how `add_method_tracer` works in general to continue supporting long-term.

### CODE REVIEW

File: lib/new_relic/agent/method_tracer.rb [hunk 1]
```diff
@@ -337,6 +326,7 @@ def _add_method_tracer_now(method_name, metric_name_code, options)
           class_eval traced_method, __FILE__, __LINE__
           alias_method _untraced_method_name(method_name, metric_name_code), method_name
           alias_method method_name, _traced_method_name(method_name, metric_name_code)
+          ruby2_keywords(_traced_method_name(method_name, metric_name_code)) if respond_to?(:ruby2_keywords, true)
```
Comment (tannalynn) [inline lib/new_relic/agent/method_tracer.rb:329]: What does `ruby2_keywords` do here?

File: test/new_relic/agent/method_tracer_params_test.rb [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+    assert_expected_results traced_class
+    refute_deprecation_warnings traced_class
+    call_expecting_warning_after_ruby_26 traced_class
+  end
+
+  [ ["untraced_methods", UntracedMethods],
+    ["traced_methods", TracedMethods],
+    ["traced_metric_methods", TracedMetricMethods],
+    ["traced_metric_methods_unscoped", TracedMetricMethodsUnscoped],
+  ].each do |traced_class_name, traced_class, metrics|
```
Comment (tannalynn) [inline test/new_relic/agent/method_tracer_params_test.rb:127]: what is `metrics` here? I'm not seeing it used below, and it seems like each will only have the first 2 things

Comment (mwlang) [reply test/new_relic/agent/method_tracer_params_test.rb:127]: need to remove `metrics`. no longer using it.

File: lib/new_relic/agent/method_tracer.rb (hunk 1 above)
Comment (mwlang) [reply lib/new_relic/agent/method_tracer.rb:329]: makes 2.7 behave like Ruby <= 2.6

File: lib/new_relic/constants.rb [hunk 3]
```diff
@@ -28,6 +28,10 @@ module NewRelic
   TRACEPARENT_KEY = "traceparent"
   TRACESTATE_KEY = "tracestate"
 
+  # Right now, old and new Rubies are the same.  Intention is to use "(...)" for
+  # Ruby 2.7+ for argument delegation, but doing so is a breaking API change.
+  ARGS_FOR_RUBY_VERSION = RUBY_VERSION < "2.7.0" ? "*args, &block" : "*args, &block"
+
```
[the remark this answers is not in the document]
Comment (mwlang) [reply lib/new_relic/constants.rb:34]: In certain scenarios, users will still see warnings, but I'm hoping the ruby2_keywords call will help guide them to the right place in their code for further inspection/resolution.

**Review by tannalynn — APPROVED** [review]
LGTM, and everything is passing. Thank you!

**Pull request merged**